### ISSUE TITLE
Compound keys with undefined values

### ISSUE THREAD

**Issue by omnight** [dumbmatter/fakeIndexedDB#15]
Hi! First, thanks for this lib, I have built a web-app that relies heavily on IndexedDB for functionality and it doesn't seem like Google would index the site properly when using the real IndexedDB, but I've had better success when using the fake lib.

My problem was with compound keys, specifically when the object I try to add did not have one or both of the key values set. To get around the problem I modified extractKey.js to handle undefined values also.

Example indata:

```
keyPath: ['parent', 'id']
value: { some_other_values: 1 }
```

In this case validateKey throws an exception since the keys have undefined value.
Here is my workaround:

```
function extractKey(keyPath, value) {
    if (Array.isArray(keyPath)) {
        var result = [];

        keyPath.forEach(function (item) {
            // This doesn't make sense to me based on the spec, but it is needed to pass the W3C KeyPath tests (see same comment in validateKey)
            if (item !== undefined && item !== null && typeof item !== 'string' && item.toString) {
                item = item.toString();
            }
		var key = extractKey(item, value);
		if(key === undefined || key === null) {
			result.push(undefined);
		} else {			
			result.push(structuredClone(validateKey(key)));
		}
        });
```

Now, this seems to work for my code at least, but I don't know if this will break specs or introduce other problems.

**Comment by dumbmatter** [comment]
Maybe I'm misunderstanding, but that will fail in real IndexedDB too, like try running this in your browser console:

    const request = indexedDB.open('test', 1);
    request.onupgradeneeded = () => {
        const db = request.result;
        const store = db.createObjectStore('store', {keyPath: ['parent', 'id']});
    };
    request.onsuccess = (event) => {
        const db = event.target.result;

        const tx = db.transaction('store', 'readwrite');

        tx.objectStore('store').put({some_other_values: 1});
    };

I think this makes sense given the spec:

[2.0 spec](https://w3c.github.io/IndexedDB/#extract-key-from-value) - below "Otherwise", step 3

[1.0 spec (what fakeIndexedDB is based on currently)](https://www.w3.org/TR/2015/REC-IndexedDB-20150108/#dfn-steps-for-extracting-a-key-from-a-value-using-a-key-path) - step 5

**Comment by omnight** [comment]
Sorry, I realize I may have over-simplified my example. The primary key is an auto-increment id and the compound key is a secondary index.

```
const request = indexedDB.open('test', 1);
request.onupgradeneeded = () => {
    const db = request.result;
    var store = db.createObjectStore('store', {keyPath: 'id', autoIncrement: true});
    store.createIndex('compound', ['a','b'], {unique: false });
};
request.onsuccess = (event) => {
    const db = event.target.result;

    const tx = db.transaction('store', 'readwrite');

    tx.objectStore('store').put({ });
    tx.objectStore('store').put({ 'a': 'x' });
    tx.objectStore('store').put({ 'a': 'y', 'b': 'z'});
}

```

**Comment by dumbmatter** [comment]
Yeah this could be a bug. In that case it shouldn't be an error, it should just not create the index entry if the keys are not all defined. I'll try to look into it soon... I think your fix would break the example in my previous post, so it'll have to be a little different than that.

**Comment by dumbmatter** [comment]
I just released v1.0.9 with a fix to this. Hope that works for you. If not, take a look at the test I wrote, maybe it's missing something still.

**Comment by omnight** [comment]
Yes, that solves my problem. Thanks!
